src/server/users/models.js

```diff
@@ -179,46 +179,84 @@ export class CreatedPostManager {
+  static getPost({isCreatedByUser, itemKey, imageKeys}) {
+    return getPromise(itemKey).then((item)=>{
+      item.isCreatedByUser = isCreatedByUser;
+      item.imageUrls = new S3Connector().getImageUrls(imageKeys);
+      return item;
+    });
+  }
+/**
+ * combine key sets for unique item key
+ * @param {Array} keySets example:
+ * [
+ *    {  entity  : 'image',
+ *       itemKey : 'item-8523910540005-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *       imageKey: 'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b000'
+ *    }
+ *    {  entity  : 'item',
+ *       itemKey : 'item-8523910540005-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *       imageKey: 'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b999'
+ *    }
+ * ]
+ * @return {Array} item example:
+ * [
+ *    {  isCreatedByUser : true,
+ *       itemKey : 'item-8523910540005-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *       imageKeys :
+ *        [
+ *          'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *          'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b999'
+ *        ]
+ *    }
+ * ]
+**/
+  static combineKeySets(keySets) {
+    const combinedKeySets = {};
+    const ret = [];
+    keySets.map((keySet)=> {
+      if (!combinedKeySets[keySet.itemKey]) {
+        // if there's no postsSet[itemKey] filed, init empty array
+        combinedKeySets[keySet.itemKey] = {
+          isCreatedByUser: 'false',
+          itemKey: keySet.itemKey,
+          imageKeys: []
+        };
       }
-      item.entity = entity;
-      item.imageUrl = new S3Connector().getImageUrl(imageKey);
-      return cb(null, item);
+      // if this item is created by this user -> isCreatedByUser = true
+      combinedKeySets[keySet.itemKey].isCreatedByUser =
+      (keySet.entity === ENTITY.ITEM) || combinedKeySets[keySet.itemKey].isCreatedByUser;
+      combinedKeySets[keySet.itemKey].imageKeys.push(keySet.imageKey);
     });
+    // convert json to array for return
+    for (const itemKey in combinedKeySets) {
+      if (combinedKeySets.hasOwnProperty(itemKey)) {
+        ret.push(combinedKeySets[itemKey]);
+      }
+    }
+    return ret;
   }
   static getPosts(userKey, cb) {
     const posts = [];
-    const promises = [];
     UserManager.getPostKeys(ENTITY.CREATED_POST, userKey)
     .then((keySets) => {
-      keySets.map((keySet)=>{
-        promises.push(new Promise((resolve, reject) => {
-          CreatedPostManager.getPost(keySet, (err, item) => {
-            if (err) {
-              reject(err);
-              return;
-            }
-            if (item.state !== STATE.REMOVED) {
-              posts.push(item);
-            }
-            resolve();
-          });
-        }));
-      });
-      Promise.all(promises).then(() => {
-        return cb(null, posts);
-      }).catch((err) => {
-        return err;
+      const combinedKeySets = CreatedPostManager.combineKeySets(keySets);
+      return Promise.all(combinedKeySets.map((combinedKeySet)=>{
```

Comment:
Array.map function returns array that is storing values returned in map func's callback block.
In this case, map's callback function return Promise objects('fetchPost').